source/common/quic/envoy_quic_server_connection.h

```diff
@@ -10,6 +11,106 @@
 namespace Envoy {
 namespace Quic {
 
+// A wrapper around the actual listener filter which checks against the given filter matcher.
+class GenericQuicListenerFilter
+    : public Network::GenericListenerFilterImplBase<Network::QuicListenerFilter> {
+public:
+  GenericQuicListenerFilter(const Network::ListenerFilterMatcherSharedPtr& matcher,
+                            Network::QuicListenerFilterPtr listener_filter)
+      : Network::GenericListenerFilterImplBase<Network::QuicListenerFilter>(
+            std::move(matcher), std::move(listener_filter)) {}
+
+  // Network::QuicListenerFilter
+  Network::FilterStatus onAccept(Network::ListenerFilterCallbacks& cb) override {
+    if (isDisabled(cb)) {
+      on_accept_by_passed_ = true;
+    }
+    return Network::GenericListenerFilterImplBase<Network::QuicListenerFilter>::onAccept(cb);
+  }
+  bool isCompatibleWithServerPreferredAddress(
+      const quic::QuicSocketAddress& server_preferred_address) const override {
+    if (on_accept_by_passed_) {
+      return true;
+    }
+    return listener_filter_->isCompatibleWithServerPreferredAddress(server_preferred_address);
+  }
+  Network::FilterStatus onPeerAddressChanged(const quic::QuicSocketAddress& new_address,
+                                             Network::Connection& connection) override {
+    if (on_accept_by_passed_) {
+      return Network::FilterStatus::Continue;
+    }
+    return listener_filter_->onPeerAddressChanged(new_address, connection);
+  }
+
+private:
+  bool on_accept_by_passed_{false};
+};
+
+// An object created on the stack during QUIC connection creation to apply listener filters, if
+// there is any, within the call stack.
+class QuicListenerFilterManagerImpl : public Network::QuicListenerFilterManager,
+                                      public Network::ListenerFilterCallbacks {
+public:
+  QuicListenerFilterManagerImpl(Event::Dispatcher& dispatcher, Network::ConnectionSocket& socket,
+                                StreamInfo::StreamInfo& stream_info)
+      : socket_(socket), stream_info_(stream_info), dispatcher_(dispatcher) {}
+
+  // Network::ListenerFilterCallbacks
+  Network::ConnectionSocket& socket() override { return socket_; }
```

Comment:
I'm a bit concerned to handle a non-const socket up.  can we add a TODO to look into use cases and see if we can make it const, or have to add docs warnings or some such?

Reply:
but for TCP if a listener filter closes the socket, it closes the single accepted TCP connection.  with UDP if we close the socket we close all QUIC connections, no?  If so I think it's dangerous to expose because folks may not be aware of behavior subtlties.  Intead  Ithink we'd need to change it to a closeConnection API which for TCP closed the socket and for QUIC reset the connection.

Reply:
No, for QUIC, closing the connection socket also means the per-connection "socket" which doesn't have an actual handle to the listening socket. Its IoHandle implementation  doesn't relay close() to the actually IoHandle object of the listening socket.